src/Processors/QueryPlan/Optimizations/distinctReadInOrder.cpp

```diff
@@ -0,0 +1,108 @@
+#include <memory>
+#include <Interpreters/ExpressionActions.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Processors/QueryPlan/DistinctStep.h>
+#include <Processors/QueryPlan/ITransformingStep.h>
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/ReadFromMergeTree.h>
+#include <Storages/ReadInOrderOptimizer.h>
+#include <Common/typeid_cast.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+size_t tryDistinctReadInOrder(QueryPlan::Node * parent_node, QueryPlan::Nodes &)
+{
+    /// check if it is preliminary distinct node
+    DistinctStep * pre_distinct = nullptr;
+    if (auto * distinct = typeid_cast<DistinctStep *>(parent_node->step.get()); distinct)
+    {
+        if (distinct->isPreliminary())
+            pre_distinct = distinct;
+    }
+    if (!pre_distinct)
+        return 0;
+
+    /// walk through the plan
+    /// check if nodes below preliminary distinct preserve sorting
+    QueryPlan::Node * node = parent_node;
+    while (!node->children.empty())
+    {
+        const auto * step = dynamic_cast<const ITransformingStep *>(node->step.get());
+        if (!step)
+            return 0;
+
+        const ITransformingStep::DataStreamTraits & traits = step->getDataStreamTraits();
+        if (!traits.preserves_sorting)
+            return 0;
+
+        node = node->children.front();
+    }
+
+    /// check if we read from MergeTree
+    auto * read_from_merge_tree = typeid_cast<ReadFromMergeTree *>(node->step.get());
+    if (!read_from_merge_tree)
+        return 0;
+
+    /// find non-const columns in DISTINCT
+    const auto & distinct_columns = pre_distinct->getOutputStream().header.getColumnsWithTypeAndName();
```

Comment:
I don't get why we need any special handling for const columns

Reply:
Please check 2 tests from this [line](https://github.com/ClickHouse/ClickHouse/pull/41014/files/eeecaf7a31e4cec82267d7369e5ee308061b10a8#diff-a6d9059ae07b9ddade618ee1ccf7ada999d07397e047c9b6d06202007c0d605bR69)